### ISSUE TITLE
"Tests: Rerun Last Run" does not work when runHandler contains async code

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode#144169]
I noticed this command doesn't work at all in the Dart extension currently, although it works fine in the TestProvider sample.

I was able to strip the Dart one down to something small which when pasted over the TestProvider sample triggers the same issue, although the issue disappears if you remove the delay at the start of `runTests`.

To repro:

- Create a new extension (with an appropriate `actiationEvent`) or open the TestProvider sample
- Paste the contents below over the contents of `extension.ts`
- Run the extension
- Switch to test side bar
- Click the run icon for the discovered "one" suite
- Note that you see a spinner for 2s, and in the Debug Console of the main VS Code window:
  ```
  Running tests!...
Test run complete!
  ```
- Now run the **Tests: Rerun Last Run** command
- Note that nothing happens

If you comment out the delay at the top of the `runTests` method, the problem goes away (my real implementation has some real async work here).

```ts
import * as vscode from 'vscode';
import * as vs from "vscode";

export async function activate(context: vscode.ExtensionContext) {
  context.subscriptions.push(new VsCodeTestController());
}

export class VsCodeTestController implements vscode.Disposable {
  private disposables: vscode.Disposable[] = [];
  public readonly controller: vs.TestController;

  constructor() {
    const controller = vs.tests.createTestController("dart", "Dart & Flutter");
    this.controller = controller;
    this.disposables.push(controller);

    controller.resolveHandler = (item: vs.TestItem | undefined) => this.resolveTestItem(item);

    controller.createRunProfile("Run", vs.TestRunProfileKind.Run, (request, token) => {
      this.runTests(false, request, token);
    });
  }

  public async runTests(debug: boolean, request: vs.TestRunRequest, token: vs.CancellationToken): Promise<void> {
    console.log("Running tests!...");
    await new Promise((resolve) => setTimeout(resolve, 10)); // REMOVING THIS LINE FIXES THE ISSUE

    const testsToRun: vs.TestItem[] = [];
    (request.include ?? this.controller.items).forEach((item) => testsToRun.push(item));

    const run = this.controller.createTestRun(request);
    for (const testToRun of testsToRun)
      run.started(testToRun);
    await new Promise((resolve) => setTimeout(resolve, 2000));
    for (const testToRun of testsToRun)
      run.passed(testToRun);
    run.end();
    console.log("Test run complete!");
  }

  private async resolveTestItem(item: vs.TestItem | undefined): Promise<void> {
    if (!item)
      await this.discover();
  }

  public async discover(): Promise<void> {
    // Discover a single fake test suite (the current file).
    await new Promise((resolve) => setTimeout(resolve, 10));
    const uri = vs.workspace.textDocuments.length ? vs.workspace.textDocuments[0].uri : vs.Uri.parse("file:///tmp/foo.dart");
    if (this.controller.items.size === 0)
      this.controller.items.add(this.controller.createTestItem("one", "one", uri));
  }

  public dispose(): any {
    this.controller.dispose();
  }
}
```

@connor4312 I tried digging through the source to see if I could understand what's going on here, but didn't get anywhere. I suspect `getTestExtIdsToRun` may not be finding anything, but I couldn't see where the results are being recorded.

**Comment by connor4312** [comment]
One note in your sample: you don't return a promise from the `runHandler` in `createRunProfile`, which causes the initial user's test run request to end without having run anything. Then, the following `createTestRun` is treated as an "external" test run, and we aren't correctly dealing with test IDs there which is causing the issue you're seeing.

**Comment by DanTup** [comment]
Ah, good spot - I'll fix that (my real extension has the same missing `return` too), and thanks for the other fix (I have some other paths that run tests outside of `runHandler`, so that will cover them too) :-)

**Comment by DanTup** [comment]
@connor4312 should it be possible to subclass `TestRunRequest` to add additional fields and have them passed back when **Tests: Rerun Last Run** is invoked? I thought I tested this earlier and it worked, but now it's not which turns out quite inconvenient 😞

The scenario is where the user invokes tests through some mechanism other than the test runner - for example, a CodeLens action. We don't have easy access to the `TestItem` to build a "normal" `TestRunRequest` so I had assumed this should work:

- Subclass `TestRunRequest` and add a new field that includes the state required to identify/run the tests triggered from CodeLens
- Construct this new `DartTestRunRequest` in the CodeLens action
- Create a new `TestRun` passing this as the `request`
- Run the tests as usual

This all works, but when the user now runs **Tests: Rerun Last Run** the `runHandler` is given a `TestRunRequest` that is not the one I provided on the last `TestRun`. This means my special state is lost, and it looks like a request to run _all_ tests.

**Comment by connor4312** [comment]
One difficulty with that is that "Rerun Last Run" works after the editor is reloaded, however if the TestRunRequest is subclassed then that is not possible. A workaround for that case would be to have some custom `data: any` property, though that isn't a pattern we follow anywhere else in the API to the best of my knowledge...

**Comment by DanTup** [comment]
Ah :-(

I believe there are many places that do provide the extension back the same instances we provide (for example I think Diagnostics given back to use in CodeAction requests, and Completions/CodeLens resolves are given back the same instance they initially returned?). None of them need to persist across reloads though.

Tbh, I'd rather have support for this than the persistance, but I'm probably biased :-)

I'll see if I can find a way to construct a normal `TestRunRequest` from CodeLens/`F5`/etc. that launch tests (it's complicated because sometimes the tests we discover statically are not the same as at runtime, for example if test names have variables in their names, are inside loops, etc.).

**Comment by connor4312** [comment]
If you're stuck needing this, please make a feature request issue for me, and I can bring it up in our next API sync 🙂

**Comment by connor4312** [comment]
@DanTup so essentially you want to add extra metadata onto the run request? I think profiles are the best way to do this. I'm a little leery of having custom metadata there as extension authors might not be aware they'll see it again if the user, for instance, reruns failed tests, which can happen both in the current session and after reload (so any custom data must also be serializable)

**Comment by DanTup** [comment]
@connor4312 

> so essentially you want to add extra metadata onto the run request?

Yep, pretty much. I already have ways of running tests, and their results show in the Test Runner, yet the Rerun Last Run command just doesn't handle them. It seems weird that I can report my test runs results into the tree, but not be able to handle a re-run.

> I'm a little leery of having custom metadata there as extension authors might not be aware they'll see it again if the user, for instance, reruns failed tests

That's specifically what this is intended to support. It could be documented on the metadata :-) I don't think there's much reason for people to use it if they're not trying to handle this - they could just skip the `TestRunRequest` entirely (as I am today) and record results if they don't want it.

> any custom data must also be serializable

That would be fine for me, it's data that's coming from `launch.json` anyway, so it's already serializable.

> I think profiles are the best way to do this

I looked at profiles in the past - I actually wanted to replace my CodeLens links with profiles - but it didn't seem like they could be scoped to specific files, and the UX wasn't as good (eg. having a menu option for "execute using profile" and then having another step to pick one was just way more clicks):

![image](https://user-images.githubusercontent.com/1078012/160405784-9a6800df-46da-441f-b4a6-d6b9cc39150d.png)

Is it possible they could be shown at the top level of that menu, and support being scoped to folders, files, or tests within?

Some examples of things this functionality is used for:

- Run Test
- Debug Test
- Run Test in Chrome (usually tests may run in the Dart VM, but for some scenarios you might want to quickly run them in the browser, where they're compiled to JS first)
- Update-goldens (run the tests, but instead of comparing the app screens to golden image files on disk, re-generate those images - for example because you've just made changes you know affect the UI)
- Run tests with arbitrary env variables set that might affect how the tests run

These things are all user-configured in `launch.json`, not provided by the extension. Users assign names to them and can choose the paths (or types or run - whole file vs specific group/tests) they're available for.

**Comment by DanTup** [comment]
@connor4312 can I reasonably use a `TestRunRequest` with a `TestItem` that isn't really in the tree?

I'm thinking about:

- stashing the run config in an object
- create a TestRunRequest that has a TestItem that has a special ID
- when asked to run tests, if the run request has an item with this special ID, fetch the config from the object

If the user reloads, we won't find the config, so we'll just run 0 tests (so if you reload and the last tests you ran were through this, they won't run - but that's already the case today, except they also don't work even if you didn't reload).

It doesn't feel like a nice fix, but I feel like I need to support this somehow. I used to have a custom command to re-run the last test session and I removed it when VS Code got its own because it seemed odd to have a "duplicate", but right now VS Code's isn't working and I don't feel I can currently move to something like profiles without the experience being much worse.
